// markov-text-generator/esm/NGram.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var NGram =
/*#__PURE__*/
function () {
  function NGram(srcWords, start, size) {
    _classCallCheck(this, NGram);

    this.srcWords = srcWords.slice(start, start + size);
  }

  _createClass(NGram, [{
    key: "wordAt",
    value: function wordAt(ind) {
      if (ind < 0 || ind >= this.srcWords.length) {
        throw new Error("Bad index in wordAt: ".concat(ind));
      }

      return this.srcWords[ind];
    }
  }, {
    key: "length",
    value: function length() {
      return this.srcWords.length;
    }
  }, {
    key: "wordsToString",
    value: function wordsToString() {
      return this.srcWords.join(" ");
    }
  }, {
    key: "shiftAdd",
    value: function shiftAdd(word) {
      var shiftedWords = this.srcWords.slice(1);
      shiftedWords.push(word);
      return new NGram(shiftedWords, 0, this.srcWords.length);
    }
  }]);

  return NGram;
}();

var _default = NGram;
exports.default = _default;
